### PR TITLE
Merge milestones/M151 to releases/shipped

### PR DESCRIPTION
**Pull request by jrbriggs**
# Release notes

- Large improvements to `git checkout` performance when there are many folder placeholders.
- A background task will now compress and deduplicate the pack-file data in the shared object cache, resulting in more sustainable disk usage.
- Improvements to the installation to more gracefully recover from an uninstallation of VFS for Git and subsequent reinstallation.
- Upgrade supports an anonymous endpoint to decouple rollouts to different engineering systems.
- Other bug fixes, engineering improvements, and test reliability fixes.

For full details, see [Milestone M151](https://github.com/Microsoft/VFSForGit/milestone/5?closed=1)

### CODE REVIEW

**Comment by derrickstolee** [comment]
I installed this version on my machine. I'll use it a bit and track the background maintenance by watching the logs.

File: GVFS/GVFS.Upgrader/UpgradeOrchestrator.cs [hunk 1]
```diff
@@ -182,8 +183,17 @@ private bool TryInitialize(out string errorMessage)
         {
             if (this.upgrader == null)
             {
+                string gitBinPath = GVFSPlatform.Instance.GitInstallation.GetInstalledGitBinPath();
+                if (string.IsNullOrEmpty(gitBinPath))
+                {
+                    errorMessage = $"nameof(this.TryInitialize): Unable to locate git installation. Ensure git is installed and try again.";
```
Comment (wilbaker) [inline GVFS/GVFS.Upgrader/UpgradeOrchestrator.cs]: `$"nameof(this.TryInitialize): ` should be `$"{nameof(this.TryInitialize)}: `

File: GVFS/GVFS.Platform.Windows/WindowsFileSystemVirtualizer.cs [hunk 2]
```diff
@@ -360,89 +252,95 @@ private HResult StartDirectoryEnumerationHandler(int commandId, Guid enumeration
...
-            BlobSizes.BlobSizesConnection blobSizesConnection,
-            int commandId,
-            Guid enumerationId,
-            string virtualPath)
+        public HResult GetDirectoryEnumerationCallback(
+        int commandId,
+        Guid enumerationId,
+        string filterFileName,
+        bool restartScan,
+        IDirectoryEnumerationResults results)
```
Comment (wilbaker) [inline GVFS/GVFS.Platform.Windows/WindowsFileSystemVirtualizer.cs]: Nit: These parameters should be indented another level

Comment (wilbaker) [reply GVFS/GVFS.Platform.Windows/WindowsFileSystemVirtualizer.cs]: It looks like several of the updated callbacks in this file have this issue.

File: GVFS/GVFS.Hooks/Program.cs [hunk 3]
```diff
@@ -113,7 +113,8 @@ private static void RemindUpgradeAvailable()
             int reminderFrequency = 10;
             int randomValue = random.Next(0, 100);
 
-            if (randomValue <= reminderFrequency && ProductUpgraderInfo.IsLocalUpgradeAvailable(tracer: null))
+            if (randomValue <= reminderFrequency &&
+                ProductUpgraderInfo.IsLocalUpgradeAvailable(tracer: null, gvfsDataRoot: GVFSHooksPlatform.GetDataRootForGVFS()))
```
Comment (kewillford) [inline GVFS/GVFS.Hooks/Program.cs]: I did a search and this method is only called here and is passing `null` for the tracer so it really doesn't need to take the tracer if it is never going to be non-null

**Pull request merged**